Error while checking build/O2/fullCI for 3c71ea91ea7c492917ab625eb1cd32d677b1d8a1 at 2023-02-25 22:57:
```
## sw/BUILD/o2checkcode-latest/log
/sw/SOURCES/O2/10773-slc8_x86-64/0/GPU/Common/test/testGPUsortCUDA.cu:22:10: error: 'boost/test/unit_test.hpp' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10773-slc8_x86-64/0/GPU/GPUTracking/TRDTracking/GPUTRDTracker.cxx:37:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10773-slc8_x86-64/0/GPU/GPUTracking/Base/GPUReconstruction.cxx:37:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10773-slc8_x86-64/0/GPU/GPUTracking/Base/GPUReconstructionCPU.cxx:45:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10773-slc8_x86-64/0/GPU/GPUTracking/display/GPUDisplay.cxx:36:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10773-slc8_x86-64/0/GPU/GPUTracking/Base/cuda/GPUReconstructionCUDAGenRTC.cu:16:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10773-slc8_x86-64/0/GPU/GPUbenchmark/cuda/../Shared/Utils.h:26:10: error: 'boost/program_options.hpp' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10773-slc8_x86-64/0/GPU/GPUbenchmark/cuda/../Shared/Utils.h:26:10: error: 'boost/program_options.hpp' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/10773-slc8_x86-64/0/Framework/Logger/include/Framework/Logger.h:14:10: error: 'fairlogger/Logger.h' file not found [clang-diagnostic-error]
/usr/local/cuda-12.0/include/cuda/std/detail/libcxx/include/type_traits:520:12: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-12.0/include/cuda/std/detail/libcxx/include/type_traits:1059:69: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-12.0/include/cuda/std/detail/libcxx/include/type_traits:1175:5: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-12.0/include/cuda/std/detail/libcxx/include/type_traits:1852:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-12.0/include/cuda/std/detail/libcxx/include/type_traits:2276:18: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-12.0/include/cuda/std/detail/libcxx/include/type_traits:4336:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-12.0/include/sm_20_atomic_functions.h:103:39: error: redefinition of 'atomicAdd' [clang-diagnostic-error]
/sw/SOURCES/O2/10773-slc8_x86-64/0/Framework/Logger/include/Framework/Logger.h:14:10: error: 'fairlogger/Logger.h' file not found [clang-diagnostic-error]
/usr/local/cuda-12.0/include/cuda/std/detail/libcxx/include/type_traits:520:12: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-12.0/include/cuda/std/detail/libcxx/include/type_traits:1059:69: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-12.0/include/cuda/std/detail/libcxx/include/type_traits:1175:5: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-12.0/include/cuda/std/detail/libcxx/include/type_traits:1852:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-12.0/include/cuda/std/detail/libcxx/include/type_traits:2276:18: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-12.0/include/cuda/std/detail/libcxx/include/type_traits:4336:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-12.0/include/sm_20_atomic_functions.h:103:39: error: redefinition of 'atomicAdd' [clang-diagnostic-error]
/sw/SOURCES/O2/10773-slc8_x86-64/0/GPU/GPUbenchmark/hip/../Shared/Utils.h:146:34: error: use of undeclared identifier 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/10773-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:8: error: unknown type name '__host__' [clang-diagnostic-error]
/sw/SOURCES/O2/10773-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:27: error: expected ';' after top level declarator [clang-diagnostic-error]
/sw/SOURCES/O2/10773-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:33: error: overloaded 'operator+=' must have at least one parameter of class or enumeration type [clang-diagnostic-error]
/sw/SOURCES/O2/10773-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:44: error: unknown type name 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/10773-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:53: error: unknown type name 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/10773-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:66:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10773-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:66:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10773-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:67:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10773-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:67:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10773-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:68:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10773-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:68:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10773-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:69:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10773-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:69:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/10773-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:85:1: error: unknown type name '__global__' [clang-diagnostic-error]
/sw/SOURCES/O2/10773-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:19: error: use of undeclared identifier 'blockIdx' [clang-diagnostic-error]
/sw/SOURCES/O2/10773-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:32: error: use of undeclared identifier 'blockDim' [clang-diagnostic-error]
/sw/SOURCES/O2/10773-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:45: error: use of undeclared identifier 'threadIdx' [clang-diagnostic-error]
/sw/SOURCES/O2/10773-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:78: error: use of undeclared identifier 'blockDim' [clang-diagnostic-error]
/sw/SOURCES/O2/10773-slc8_x86-64/0/GPU/GPUbenchmark/hip/../Shared/Utils.h:146:34: error: use of undeclared identifier 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/10773-slc8_x86-64/0/GPU/GPUbenchmark/hip/benchmark.hip.cxx:199:35: error: use of undeclared identifier 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/slc8_x86-64/GCC-Toolchain/v12.2.0-alice1-2/lib/gcc/x86_64-unknown-linux-gnu/12.2.0/../../../../include/c++/12.2.0/bits/shared_ptr_base.h:196:22: error: expected expression [clang-diagnostic-error]
/sw/slc8_x86-64/GCC-Toolchain/v12.2.0-alice1-2/lib/gcc/x86_64-unknown-linux-gnu/12.2.0/../../../../include/c++/12.2.0/bits/shared_ptr_base.h:196:22: error: type name does not allow function specifier to be specified [clang-diagnostic-error]
/sw/slc8_x86-64/GCC-Toolchain/v12.2.0-alice1-2/lib/gcc/x86_64-unknown-linux-gnu/12.2.0/../../../../include/c++/12.2.0/bits/shared_ptr_base.h:196:22: error: use of undeclared identifier 'noinline'; did you mean 'inline'? [clang-diagnostic-error]
/sw/SOURCES/O2/10773-slc8_x86-64/0/Detectors/EMCAL/calibration/include/EMCALCalibration/EMCALCalibExtractor.h:36:10: error: 'omp.h' file not found [clang-diagnostic-error]
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/10773/3c71ea91ea7c492917ab625eb1cd32d677b1d8a1/build_O2_fullCI/pretty.html).

--- a/Detectors/MUON/MID/Filtering/CMakeLists.txt
+++ b/Detectors/MUON/MID/Filtering/CMakeLists.txt
@@ -12,8 +12,11 @@
 o2_add_library(
   MIDFiltering
   SOURCES src/ChannelMasksHandler.cxx src/ChannelScalers.cxx src/FetToDead.cxx
-          src/MaskMaker.cxx src/FiltererBC.cxx
+          src/MaskMaker.cxx src/FiltererBC.cxx src/FiltererBCParam.cxx
   PUBLIC_LINK_LIBRARIES O2::CommonDataFormat O2::DataFormatsMID O2::MIDBase O2::MIDRaw Microsoft.GSL::GSL)
+
+o2_target_root_dictionary(MIDFiltering
+  HEADERS include/MIDFiltering/FiltererBCParam.h)
 
 add_subdirectory(exe)
 

--- a/Detectors/MUON/MID/Filtering/CMakeLists.txt
+++ b/Detectors/MUON/MID/Filtering/CMakeLists.txt
@@ -1,19 +1,19 @@
-# Copyright CERN and copyright holders of ALICE O2. This software is distributed
-# under the terms of the GNU General Public License v3 (GPL Version 3), copied
-# verbatim in the file "COPYING".
+# Copyright 2019-2020 CERN and copyright holders of ALICE O2.
+# See https://alice-o2.web.cern.ch/copyright for details of the copyright holders.
+# All rights not expressly granted are reserved.
 #
-# See http://alice-o2.web.cern.ch/license for full licensing information.
+# This software is distributed under the terms of the GNU General Public
+# License v3 (GPL Version 3), copied verbatim in the file "COPYING".
 #
 # In applying this license CERN does not waive the privileges and immunities
-# granted to it by virtue of its status as an Intergovernmental Organization or
-# submit itself to any jurisdiction.
+# granted to it by virtue of its status as an Intergovernmental Organization
+# or submit itself to any jurisdiction.
 
 o2_add_library(
   MIDFiltering
   SOURCES src/ChannelMasksHandler.cxx src/ChannelScalers.cxx src/FetToDead.cxx
-          src/MaskMaker.cxx
-  PUBLIC_LINK_LIBRARIES O2::DataFormatsMID O2::MIDBase O2::MIDRaw
-                        Microsoft.GSL::GSL)
+          src/MaskMaker.cxx src/FiltererBC.cxx
+  PUBLIC_LINK_LIBRARIES O2::CommonDataFormat O2::DataFormatsMID O2::MIDBase O2::MIDRaw Microsoft.GSL::GSL)
 
 add_subdirectory(exe)
 

--- a/Detectors/MUON/MID/Filtering/exe/CMakeLists.txt
+++ b/Detectors/MUON/MID/Filtering/exe/CMakeLists.txt
@@ -1,12 +1,14 @@
-# Copyright CERN and copyright holders of ALICE O2. This software is distributed
-# under the terms of the GNU General Public License v3 (GPL Version 3), copied
-# verbatim in the file "COPYING".
+# Copyright 2019-2020 CERN and copyright holders of ALICE O2.
+# See https://alice-o2.web.cern.ch/copyright for details of the copyright holders.
+# All rights not expressly granted are reserved.
 #
-# See http://alice-o2.web.cern.ch/license for full licensing information.
+# This software is distributed under the terms of the GNU General Public
+# License v3 (GPL Version 3), copied verbatim in the file "COPYING".
 #
 # In applying this license CERN does not waive the privileges and immunities
-# granted to it by virtue of its status as an Intergovernmental Organization or
-# submit itself to any jurisdiction.
+# granted to it by virtue of its status as an Intergovernmental Organization
+# or submit itself to any jurisdiction.
+
 
 o2_add_executable(
   mask-maker

--- a/Detectors/MUON/MID/Filtering/include/MIDFiltering/FiltererBC.h
+++ b/Detectors/MUON/MID/Filtering/include/MIDFiltering/FiltererBC.h
@@ -1,0 +1,68 @@
+// Copyright 2019-2020 CERN and copyright holders of ALICE O2.
+// See https://alice-o2.web.cern.ch/copyright for details of the copyright holders.
+// All rights not expressly granted are reserved.
+//
+// This software is distributed under the terms of the GNU General Public
+// License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// \file   MIDFiltering/FiltererBC.h
+/// \brief  BC filterer for MID
+/// \author Diego Stocco <Diego.Stocco at cern.ch>
+/// \date   23 January 2023
+
+#ifndef O2_MID_FILTERERBC_H
+#define O2_MID_FILTERERBC_H
+
+#include <vector>
+#include <gsl/gsl>
+#include "CommonDataFormat/BunchFilling.h"
+#include "DataFormatsMID/ROFRecord.h"
+
+namespace o2
+{
+namespace mid
+{
+/// Filtering algorithm for MID
+class FiltererBC
+{
+ public:
+  /// @brief Filters the data BC
+  /// @param rofRecords ROF records
+  /// @returns Vector of filtered ROF records
+  std::vector<ROFRecord> process(gsl::span<const ROFRecord> rofRecords);
+
+  /// @brief Set the maximum BC diff in the lower side
+  /// @param bcDiffLow maximum BC diff in the lower side (negative value)
+  void setBCDiffLow(int bcDiffLow) { mBCDiffLow = bcDiffLow; }
+
+  /// @brief Set the maximum BC diff in the upper side
+  /// @param bcDiffHigh maximum BC diff in the upper side (positive value)
+  void setBCDiffHigh(int bcDiffHigh) { mBCDiffHigh = bcDiffHigh; }
+
+  /// @brief Only selects BCs but do not merge them
+  /// @param selectOnly Flag to only select BCs without merging events
+  void setSelectOnly(bool selectOnly) { mSelectOnly = selectOnly; }
+
+  /// @brief Sets the bunch filling scheme
+  /// @param bunchFilling Bunch filling scheme
+  void setBunchFilling(const BunchFilling& bunchFilling) { mBunchFilling = bunchFilling; }
+
+ private:
+  /// @brief Returns the matched collision BC
+  /// @param bc BC to be checked
+  /// @return The BC of the matched collision or -1 if there is no match
+  int matchedCollision(int bc);
+
+  int mBCDiffLow = 0;         /// Maximum BC diff on the lower side
+  int mBCDiffHigh = 0;        /// Maximum BC diff on the higher side
+  bool mSelectOnly = false;   /// Flag to only select BCs without merging digits
+  BunchFilling mBunchFilling; /// Bunch filling scheme
+};
+} // namespace mid
+} // namespace o2
+
+#endif /* O2_MID_FILTERERBC_H */

--- a/Detectors/MUON/MID/Filtering/include/MIDFiltering/FiltererBCParam.h
+++ b/Detectors/MUON/MID/Filtering/include/MIDFiltering/FiltererBCParam.h
@@ -1,0 +1,38 @@
+// Copyright 2019-2020 CERN and copyright holders of ALICE O2.
+// See https://alice-o2.web.cern.ch/copyright for details of the copyright holders.
+// All rights not expressly granted are reserved.
+//
+// This software is distributed under the terms of the GNU General Public
+// License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+#ifndef O2_MID_FILTERERBCPARAM_H
+#define O2_MID_FILTERERBCPARAM_H
+
+#include "CommonUtils/ConfigurableParam.h"
+#include "CommonUtils/ConfigurableParamHelper.h"
+
+namespace o2
+{
+namespace mid
+{
+
+/**
+ * @class FiltererBCParam
+ * @brief Configurable parameters for the BC filterer
+ */
+struct FiltererBCParam : public o2::conf::ConfigurableParamHelper<FiltererBCParam> {
+
+  int maxBCDiffLow = -1;   ///< Maximum BC diff in the lower side
+  int maxBCDiffHigh = 1;   ///< Maximum BC diff in the upper side
+  bool selectOnly = false; ///< Selects BCs but does not merge them
+
+  O2ParamDef(FiltererBCParam, "MIDFiltererBC");
+};
+} // namespace mid
+} // namespace o2
+
+#endif

--- a/Detectors/MUON/MID/Filtering/src/FiltererBC.cxx
+++ b/Detectors/MUON/MID/Filtering/src/FiltererBC.cxx
@@ -1,0 +1,113 @@
+// Copyright 2019-2020 CERN and copyright holders of ALICE O2.
+// See https://alice-o2.web.cern.ch/copyright for details of the copyright holders.
+// All rights not expressly granted are reserved.
+//
+// This software is distributed under the terms of the GNU General Public
+// License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// \file   MID/Filtering/src/FiltererBC.cxx
+/// \brief  BC filterer for MID
+/// \author Diego Stocco <Diego.Stocco at cern.ch>
+/// \date   23 January 2023
+
+#include "MIDFiltering/FiltererBC.h"
+
+namespace o2
+{
+namespace mid
+{
+
+int FiltererBC::matchedCollision(int bc)
+{
+  // First we check the simple case to save time
+  if (mBunchFilling.testInteractingBC(bc)) {
+    return bc;
+  }
+
+  // Then we check for neighbor BCs
+  // We split the check in two so that we start the check from the input bc
+  // and we move farther from it.
+
+  int matchedBC = std::numeric_limits<int>::max();
+
+  // We start on the upper direction
+
+  // The allowed BC are o2::constants::lhc::LHCMaxBunches.
+  // Numeration starts from 0, so o2::constants::lhc::LHCMaxBunches is not allowed.
+  // This ensures that we do not check out of boundaries
+  int end = bc + mBCDiffHigh;
+  if (end >= o2::constants::lhc::LHCMaxBunches) {
+    end = o2::constants::lhc::LHCMaxBunches - 1;
+  }
+  // We exclude the input BC since it was already checked before
+  for (int ibc = bc + 1; ibc <= end; ++ibc) {
+    if (mBunchFilling.testInteractingBC(ibc)) {
+      // We do not stop here because we want first to check that there is no closer matching in the other direction
+      matchedBC = ibc;
+      break;
+    }
+  }
+
+  // We then check the lower direction
+  end = bc + mBCDiffLow;
+  if (end < 0) {
+    end = 0;
+  }
+  // We exclude the input BC since it was already checked before
+  for (int ibc = bc - 1; ibc >= end; --ibc) {
+    if (mBunchFilling.testInteractingBC(ibc)) {
+      if (bc - ibc < matchedBC - bc) {
+        // This collision BC is closer to the input BC than the one found in the upper side
+        return ibc;
+      }
+      break;
+    }
+  }
+
+  return matchedBC;
+}
+
+std::vector<ROFRecord> FiltererBC::process(gsl::span<const ROFRecord> rofRecords)
+{
+  std::vector<ROFRecord> filteredROFs;
+  auto rofIt = rofRecords.begin();
+  auto end = rofRecords.end();
+  // Loop on ROFs
+  for (; rofIt != end; ++rofIt) {
+    // Check if BC matches a collision BC
+    auto matchedColl = matchedCollision(rofIt->interactionRecord.bc);
+    if (matchedColl < o2::constants::lhc::LHCMaxBunches) {
+      // Add this ROF to the filtered ones
+      filteredROFs.emplace_back(*rofIt);
+      if (mSelectOnly) {
+        continue;
+      }
+      // Use the BC of the matching collision
+      filteredROFs.back().interactionRecord.bc = matchedColl;
+      // Search for neighbor BCs
+      for (auto auxIt = rofIt + 1; auxIt != end; ++auxIt) {
+        // We assume that data are time-ordered
+        if (auxIt->interactionRecord.orbit != rofIt->interactionRecord.orbit) {
+          // Orbit does not match
+          break;
+        }
+        int bcDiff = auxIt->interactionRecord.bc - matchedColl;
+        if (bcDiff < mBCDiffLow || bcDiff > mBCDiffHigh) {
+          // BC is not in the allowed window
+          break;
+        }
+        filteredROFs.back().nEntries += auxIt->nEntries;
+        // CAVEAT: we are updating rofIt here. Do not use it in the following
+        rofIt = auxIt;
+      }
+    }
+  }
+  return filteredROFs;
+}
+
+} // namespace mid
+} // namespace o2

--- a/Detectors/MUON/MID/Filtering/src/FiltererBCParam.cxx
+++ b/Detectors/MUON/MID/Filtering/src/FiltererBCParam.cxx
@@ -1,0 +1,14 @@
+// Copyright 2019-2020 CERN and copyright holders of ALICE O2.
+// See https://alice-o2.web.cern.ch/copyright for details of the copyright holders.
+// All rights not expressly granted are reserved.
+//
+// This software is distributed under the terms of the GNU General Public
+// License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+#include "MIDFiltering/FiltererBCParam.h"
+
+O2ParamImpl(o2::mid::FiltererBCParam);

--- a/Detectors/MUON/MID/Filtering/src/MIDFilteringLinkDef.h
+++ b/Detectors/MUON/MID/Filtering/src/MIDFilteringLinkDef.h
@@ -1,0 +1,20 @@
+// Copyright 2019-2020 CERN and copyright holders of ALICE O2.
+// See https://alice-o2.web.cern.ch/copyright for details of the copyright holders.
+// All rights not expressly granted are reserved.
+//
+// This software is distributed under the terms of the GNU General Public
+// License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+#ifdef __CLING__
+
+#pragma link off all globals;
+#pragma link off all classes;
+#pragma link off all functions;
+
+#pragma link C++ class o2::mid::FiltererBCParam + ;
+
+#endif

--- a/Detectors/MUON/MID/Filtering/test/CMakeLists.txt
+++ b/Detectors/MUON/MID/Filtering/test/CMakeLists.txt
@@ -1,12 +1,13 @@
-# Copyright CERN and copyright holders of ALICE O2. This software is distributed
-# under the terms of the GNU General Public License v3 (GPL Version 3), copied
-# verbatim in the file "COPYING".
+# Copyright 2019-2020 CERN and copyright holders of ALICE O2.
+# See https://alice-o2.web.cern.ch/copyright for details of the copyright holders.
+# All rights not expressly granted are reserved.
 #
-# See http://alice-o2.web.cern.ch/license for full licensing information.
+# This software is distributed under the terms of the GNU General Public
+# License v3 (GPL Version 3), copied verbatim in the file "COPYING".
 #
 # In applying this license CERN does not waive the privileges and immunities
-# granted to it by virtue of its status as an Intergovernmental Organization or
-# submit itself to any jurisdiction.
+# granted to it by virtue of its status as an Intergovernmental Organization
+# or submit itself to any jurisdiction.
 
 o2_add_test(
   Filtering

--- a/Detectors/MUON/MID/Filtering/test/testFilter.cxx
+++ b/Detectors/MUON/MID/Filtering/test/testFilter.cxx
@@ -29,6 +29,7 @@
 #include "MIDFiltering/ChannelScalers.h"
 #include "MIDFiltering/FetToDead.h"
 #include "MIDFiltering/MaskMaker.h"
+#include "MIDFiltering/FiltererBC.h"
 
 namespace o2
 {
@@ -44,7 +45,7 @@ std::vector<ColumnData> generateData(size_t nData = 10)
   std::uniform_int_distribution<uint8_t> colIds(0, 6);
   std::uniform_int_distribution<uint16_t> patterns(0, 0xFFFF);
 
-  std::vector<o2::mid::ColumnData> data;
+  std::vector<ColumnData> data;
   for (size_t idata = 0; idata < nData; ++idata) {
     ColumnData col;
     col.deId = deIds(mt);
@@ -60,7 +61,7 @@ std::vector<ColumnData> generateData(size_t nData = 10)
 
 BOOST_AUTO_TEST_CASE(mask)
 {
-  o2::mid::ColumnData col1;
+  ColumnData col1;
   col1.deId = 71;
   col1.columnId = 6;
   col1.setNonBendPattern(0x8000);
@@ -80,12 +81,12 @@ BOOST_AUTO_TEST_CASE(mask)
 BOOST_AUTO_TEST_CASE(scalers)
 {
 
-  o2::mid::ColumnData col1;
+  ColumnData col1;
   col1.deId = 71;
   col1.columnId = 6;
   col1.setNonBendPattern(0x8000);
 
-  o2::mid::ChannelScalers cs;
+  ChannelScalers cs;
   cs.count(col1);
   auto sc1 = cs.getScalers();
   BOOST_REQUIRE(sc1.size() == 1);
@@ -98,7 +99,7 @@ BOOST_AUTO_TEST_CASE(scalers)
   }
   cs.reset();
 
-  o2::mid::ColumnData col2;
+  ColumnData col2;
   col2.deId = 25;
   col2.columnId = 3;
   col2.setBendPattern(0x0100, 3);
@@ -117,12 +118,12 @@ BOOST_AUTO_TEST_CASE(scalers)
 BOOST_AUTO_TEST_CASE(maskMaker)
 {
   auto data = generateData();
-  o2::mid::ChannelScalers cs;
-  std::vector<o2::mid::ColumnData> refMasks{};
+  ChannelScalers cs;
+  std::vector<ColumnData> refMasks{};
   for (auto col : data) {
     cs.reset();
     cs.count(col);
-    auto masks = o2::mid::makeMasks(cs, 1, 0., refMasks);
+    auto masks = makeMasks(cs, 1, 0., refMasks);
     BOOST_TEST(masks.size() == 1);
     for (auto& mask : masks) {
       for (int iline = 0; iline < 4; ++iline) {
@@ -166,6 +167,55 @@ BOOST_AUTO_TEST_CASE(FETConversion)
   // Since the mask in the FET conversion takes care of removing it
   BOOST_TEST(inverted.back().getBendPattern(3) == 0);
   BOOST_TEST(inverted.back().getNonBendPattern() == 0);
+}
+
+BOOST_AUTO_TEST_CASE(filterBC)
+{
+  /// Tests the BC filtering
+  FiltererBC filterBC;
+  BunchFilling bcFill;
+  int collBC1 = 100;
+  int collBC2 = 105;
+  int bcDiffLow = -1;
+  int bcDiffHigh = 1;
+  bcFill.setBC(collBC1);
+  bcFill.setBC(collBC2);
+
+  filterBC.setBunchFilling(bcFill);
+  filterBC.setBCDiffLow(bcDiffLow);
+  filterBC.setBCDiffHigh(bcDiffHigh);
+  std::vector<ROFRecord> rofs;
+  int nColBC1_1 = 1;
+  int nColBC1_2 = 2;
+  int nColBC2 = 4;
+
+  // Data compatible with collision BC1
+  InteractionRecord ir(collBC1 + bcDiffLow, 1);
+  rofs.emplace_back(ir, EventType::Standard, 0, nColBC1_1);
+
+  // Data compatible with collision BC1
+  ir.bc = collBC1 + bcDiffHigh;
+  rofs.emplace_back(ir, EventType::Standard, rofs.back().getEndIndex(), nColBC1_2);
+
+  // Data not compatible with collision BC
+  ir.bc = collBC1 + bcDiffHigh + 1;
+  rofs.emplace_back(ir, EventType::Standard, rofs.back().getEndIndex(), 1);
+
+  // Data compatible with collision BC2
+  ir.bc = collBC2;
+  rofs.emplace_back(ir, EventType::Standard, rofs.back().getEndIndex(), nColBC2);
+
+  auto filteredROFs = filterBC.process(rofs);
+
+  BOOST_REQUIRE(filteredROFs.size() == 2);
+
+  // Check that the first two are merged
+  BOOST_TEST(filteredROFs.front().interactionRecord.bc == collBC1);
+  BOOST_TEST(filteredROFs.front().nEntries == nColBC1_1 + nColBC1_2);
+
+  // Check that the last is kept unchanged
+  BOOST_TEST(filteredROFs.back().interactionRecord.bc == collBC2);
+  BOOST_TEST(filteredROFs.back().nEntries == nColBC2);
 }
 
 } // namespace mid

--- a/Detectors/MUON/MID/Workflow/CMakeLists.txt
+++ b/Detectors/MUON/MID/Workflow/CMakeLists.txt
@@ -1,13 +1,13 @@
-# Copyright 2019-2020 CERN and copyright holders of ALICE O2. See
-# https://alice-o2.web.cern.ch/copyright for details of the copyright holders.
+# Copyright 2019-2022 CERN and copyright holders of ALICE O2.
+# See https://alice-o2.web.cern.ch/copyright for details of the copyright holders.
 # All rights not expressly granted are reserved.
 #
-# This software is distributed under the terms of the GNU General Public License
-# v3 (GPL Version 3), copied verbatim in the file "COPYING".
+# This software is distributed under the terms of the GNU General Public
+# License v3 (GPL Version 3), copied verbatim in the file "COPYING".
 #
 # In applying this license CERN does not waive the privileges and immunities
-# granted to it by virtue of its status as an Intergovernmental Organization or
-# submit itself to any jurisdiction.
+# granted to it by virtue of its status as an Intergovernmental Organization
+# or submit itself to any jurisdiction.
 
 o2_add_library(
   MIDWorkflow
@@ -21,6 +21,7 @@ o2_add_library(
           src/EntropyDecoderSpec.cxx
           src/EntropyEncoderSpec.cxx
           src/FilteringSpec.cxx
+          src/FilteringBCSpec.cxx
           src/MaskHandlerSpec.cxx
           src/MaskMakerSpec.cxx
           src/ChannelCalibratorSpec.cxx

--- a/Detectors/MUON/MID/Workflow/README.md
+++ b/Detectors/MUON/MID/Workflow/README.md
@@ -117,6 +117,32 @@ o2-mid-reco-workflow --change-local-to-BC <value>
 
 where `<value>` is the chosen offset in number of BCs (can be negative).
 
+## MID BC filtering
+
+The MID time resolution is better than 25 ns, allowing it to distinguish between different BCs.
+However, a spread of the signal was observed, probably due an insufficient equalization of the delays across the various detection element.
+In order to study and possibly correct for any inefficiency arising from this spread, it is possible to select the BCs corresponding to a collision and merge the digits in a configurable window around this BC.
+To this aim, it is enough to run the reconstruction with the option:
+
+```bash
+o2-mid-reco-workflow --enable-filter-BC
+```
+
+The time window can be tuned with:
+
+```bash
+--configKeyValues="MIDFiltererBC.maxBCDiffLow=-1;MIDFiltererBC.maxBCDiffHigh=1"
+```
+
+Notice that the `maxBCDiffLow` has to be a negative value.
+
+It is also possible to only select the collision BC, without merging the digits in the corresponding window.
+This can be done adding the option:
+
+```bash
+--configKeyValues="MIDFiltererBC.selectOnly=1"
+```
+
 ### Reconstruction options
 
 By default, the reconstruction produces clusters and tracks that are written on file.

--- a/Detectors/MUON/MID/Workflow/include/MIDWorkflow/FilteringBCSpec.h
+++ b/Detectors/MUON/MID/Workflow/include/MIDWorkflow/FilteringBCSpec.h
@@ -1,0 +1,32 @@
+// Copyright 2019-2020 CERN and copyright holders of ALICE O2.
+// See https://alice-o2.web.cern.ch/copyright for details of the copyright holders.
+// All rights not expressly granted are reserved.
+//
+// This software is distributed under the terms of the GNU General Public
+// License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// \file   MIDWorkflow/FilteringBCSpec.h
+/// \brief  MID filtering spec
+/// \author Diego Stocco <Diego.Stocco at cern.ch>
+/// \date   16 March 2022
+
+#ifndef O2_MID_FILTERINGBCSPEC_H
+#define O2_MID_FILTERINGBCSPEC_H
+
+#include "Framework/DataProcessorSpec.h"
+
+#include <string_view>
+
+namespace o2
+{
+namespace mid
+{
+framework::DataProcessorSpec getFilteringBCSpec(bool useMC = true, std::string_view inDesc = "FDATA");
+}
+} // namespace o2
+
+#endif // O2_MID_FILTERINGBCSPEC_H

--- a/Detectors/MUON/MID/Workflow/src/FilteringBCSpec.cxx
+++ b/Detectors/MUON/MID/Workflow/src/FilteringBCSpec.cxx
@@ -1,0 +1,122 @@
+// Copyright 2019-2020 CERN and copyright holders of ALICE O2.
+// See https://alice-o2.web.cern.ch/copyright for details of the copyright holders.
+// All rights not expressly granted are reserved.
+//
+// This software is distributed under the terms of the GNU General Public
+// License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// \file   MID/Workflow/src/FilteringBCSpec.cxx
+/// \brief  MID filtering spec
+/// \author Diego Stocco <Diego.Stocco at cern.ch>
+/// \date   16 March 2022
+
+#include "MIDWorkflow/FilteringBCSpec.h"
+
+#include <vector>
+#include <gsl/gsl>
+#include <fmt/format.h>
+#include "Framework/CCDBParamSpec.h"
+#include "Framework/Output.h"
+#include "Framework/Task.h"
+#include "Framework/WorkflowSpec.h"
+#include "DataFormatsMID/ColumnData.h"
+#include "DataFormatsMID/ROFRecord.h"
+#include "DataFormatsMID/MCLabel.h"
+#include "DetectorsBase/GRPGeomHelper.h"
+#include "SimulationDataFormat/MCTruthContainer.h"
+#include "MIDFiltering/FiltererBC.h"
+#include "MIDFiltering/FiltererBCParam.h"
+#include "MIDSimulation/DigitsMerger.h"
+#include "MIDWorkflow/ColumnDataSpecsUtils.h"
+
+namespace of = o2::framework;
+
+namespace o2
+{
+namespace mid
+{
+
+class FilteringBCDeviceDPL
+{
+ public:
+  FilteringBCDeviceDPL(std::shared_ptr<o2::base::GRPGeomRequest> gr, bool useMC, std::vector<of::OutputSpec> outputSpecs) : mGGCCDBRequest(gr), mUseMC(useMC)
+  {
+    mOutputs = specs::buildOutputs(outputSpecs);
+  }
+
+  void init(of::InitContext& ic)
+  {
+    o2::base::GRPGeomHelper::instance().setRequest(mGGCCDBRequest);
+    mFiltererBC.setBCDiffLow(FiltererBCParam::Instance().maxBCDiffLow);
+    mFiltererBC.setBCDiffHigh(FiltererBCParam::Instance().maxBCDiffHigh);
+    mFiltererBC.setSelectOnly(FiltererBCParam::Instance().selectOnly);
+  }
+
+  void finaliseCCDB(of::ConcreteDataMatcher matcher, void* obj)
+  {
+    if (o2::base::GRPGeomHelper::instance().finaliseCCDB(matcher, obj)) {
+      LOG(info) << "Update Bunch Filling for MID";
+      mFiltererBC.setBunchFilling(o2::base::GRPGeomHelper::instance().getGRPLHCIF()->getBunchFilling());
+    }
+  }
+
+  void run(of::ProcessingContext& pc)
+  {
+    updateTimeDependentParams(pc);
+
+    auto data = specs::getData(pc, "mid_filter_BC_in", EventType::Standard);
+    auto inROFRecords = specs::getRofs(pc, "mid_filter_BC_in", EventType::Standard);
+
+    auto inMCContainer = mUseMC ? specs::getLabels(pc, "mid_filter_BC_in") : nullptr;
+
+    auto filteredROFs = mFiltererBC.process(inROFRecords);
+    mDigitsMerger.process(data, filteredROFs, inMCContainer.get());
+
+    pc.outputs().snapshot(mOutputs[0], mDigitsMerger.getColumnData());
+    pc.outputs().snapshot(mOutputs[1], mDigitsMerger.getROFRecords());
+    if (mUseMC) {
+      pc.outputs().snapshot(mOutputs[2], mDigitsMerger.getMCContainer());
+    }
+  }
+
+ private:
+  void updateTimeDependentParams(o2::framework::ProcessingContext& pc)
+  {
+    // Triggers finalizeCCDB
+    o2::base::GRPGeomHelper::instance().checkUpdates(pc);
+  }
+
+  std::shared_ptr<o2::base::GRPGeomRequest> mGGCCDBRequest;
+  bool mUseMC{false};
+  FiltererBC mFiltererBC;
+  DigitsMerger mDigitsMerger;
+  std::vector<of::Output> mOutputs;
+};
+
+of::DataProcessorSpec getFilteringBCSpec(bool useMC, std::string_view inDesc)
+{
+
+  auto inputSpecs = specs::buildInputSpecs("mid_filter_BC_in", inDesc, useMC);
+  auto ggRequest = std::make_shared<o2::base::GRPGeomRequest>(false,                          // orbitResetTime
+                                                              false,                          // GRPECS=true
+                                                              true,                           // GRPLHCIF
+                                                              false,                          // GRPMagField
+                                                              false,                          // askMatLUT
+                                                              o2::base::GRPGeomRequest::None, // geometry
+                                                              inputSpecs,
+                                                              false);
+
+  auto outputSpecs = specs::buildStandardOutputSpecs("mid_filter_BC_out", "BDATA", useMC);
+
+  return of::DataProcessorSpec{
+    "MIDFilteringBC",
+    {inputSpecs},
+    {outputSpecs},
+    of::AlgorithmSpec{of::adaptFromTask<o2::mid::FilteringBCDeviceDPL>(ggRequest, useMC, outputSpecs)}};
+}
+} // namespace mid
+} // namespace o2

--- a/Detectors/MUON/MID/Workflow/src/reco-workflow.cxx
+++ b/Detectors/MUON/MID/Workflow/src/reco-workflow.cxx
@@ -29,6 +29,7 @@
 #include "DataFormatsMID/MCClusterLabel.h"
 #include "MIDWorkflow/ClusterizerSpec.h"
 #include "MIDWorkflow/FilteringSpec.h"
+#include "MIDWorkflow/FilteringBCSpec.h"
 #include "MIDWorkflow/TimingSpec.h"
 #include "MIDWorkflow/TrackerSpec.h"
 #include "CommonUtils/ConfigurableParam.h"
@@ -51,6 +52,7 @@ void customize(std::vector<ConfigParamSpec>& workflowOptions)
       {"disable-root-output", VariantType::Bool, false, {"Do not write output to file"}},
       {"change-local-to-BC", VariantType::Int, 0, {"Change the delay between the MID local clock and the BC"}},
       {"disable-filtering", VariantType::Bool, false, {"Do not remove bad channels"}},
+      {"enable-filter-BC", VariantType::Bool, false, {"Filter BCs"}},
       {"configKeyValues", VariantType::String, "", {"Semicolon separated key=value strings"}}};
 
   workflowOptions.insert(workflowOptions.end(), options.begin(), options.end());
@@ -65,22 +67,35 @@ WorkflowSpec defineDataProcessing(ConfigContext const& cfgc)
   bool disableFile = cfgc.options().get<bool>("disable-root-output");
   auto disableFiltering = cfgc.options().get<bool>("disable-filtering");
   auto localToBC = cfgc.options().get<int>("change-local-to-BC");
+  auto filterBC = cfgc.options().get<bool>("enable-filter-BC");
   o2::conf::ConfigurableParam::updateFromString(cfgc.options().get<std::string>("configKeyValues"));
 
   WorkflowSpec specs;
-  std::string dataDesc = "DATA";
-  std::string rofDesc = "DATAROF";
-  std::string labelsDesc = "DATALABELS";
+  std::string baseDataDesc = "DATA";
+  std::string baseROFDesc = "DATAROF";
+  std::string baseLabelsDesc = "DATALABELS";
+  std::string dataDesc = baseDataDesc;
+  std::string rofDesc = baseROFDesc;
+  std::string labelsDesc = baseLabelsDesc;
   if (!disableFiltering) {
-    std::string outDataDesc = "F" + dataDesc;
+    std::string prefix = "F";
+    std::string outDataDesc = prefix + baseDataDesc;
     specs.emplace_back(o2::mid::getFilteringSpec(!disableMC, dataDesc, outDataDesc));
     dataDesc = outDataDesc;
-    rofDesc = "F" + rofDesc;
-    labelsDesc = "F" + labelsDesc;
+    rofDesc = prefix + baseROFDesc;
+    labelsDesc = prefix + baseLabelsDesc;
   }
   if (localToBC != 0) {
     specs.emplace_back(o2::mid::getTimingSpec(localToBC, rofDesc));
-    rofDesc = "TDATAROF";
+    std::string prefix = "T";
+    rofDesc = prefix + baseROFDesc;
+  }
+  if (filterBC) {
+    specs.emplace_back(o2::mid::getFilteringBCSpec(!disableMC, dataDesc));
+    std::string prefix = "B";
+    dataDesc = prefix + baseDataDesc;
+    rofDesc = prefix + baseROFDesc;
+    labelsDesc = prefix + baseLabelsDesc;
   }
   specs.emplace_back(o2::mid::getClusterizerSpec(!disableMC, dataDesc, rofDesc, labelsDesc));
   if (!disableTracking) {

